### PR TITLE
Role Basic edit and Permission Edit

### PR DESCRIPTION
**Pull request by DimalChandrasiri**
**Basic Edit**
Screenshot 2020-03-17 at 22 08 09

**Permission Edit**
Screenshot 2020-03-17 at 22 08 25

### CODE REVIEW

File: apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    } = props;
+
+    /**
+     * Util method to check permissions already in the selected role.
+     * 
+     * @param permList - Permissions which are available in the selected role
+     * @param nodes - Permmission Node List
+     * @param isParentChecked - Check whether parent is selected
+     */
+    const checkRolePerms = (permList: string[], nodes: any, isParentChecked: boolean): void  => {
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx]: Shall we use qualified variable declarations e.g.  `checkRolePermissions` ? Noticed in other places as well

File: apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            getPermsList();
+        }
+    },[]);
+
+    /**
+     * A util method which will check all the child elements.
+     * 
+     * @param nodes child nodes need to be checked.
+     */
+    const applyCheckStateTo = (nodes: any, checked: boolean): void => {
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx]: Shall we make the declaration more meaningful here? Doesn't get the exact meaning by looking at the variable declaration name.

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx [hunk 3]
```diff
@@ -51,8 +51,8 @@ interface CreateRoleProps {
  */
 enum WizardStepsFormTypes {
     BASIC_DETAILS = "BasicDetails",
-    PERM_LIST = "PermissionList",
-    USER_LIST = "RoleUserList",
+    PERM_LIST= "PermissionList",
+    USER_LIST= "RoleUserList",
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]:
```suggestion
    PERM_LIST = "PermissionList",
    USER_LIST = "RoleUserList",
```

Why is this space removed ?

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx [hunk 4]
```diff
@@ -205,7 +132,7 @@ export const CreateRoleWizard: FunctionComponent<CreateRoleProps> = (props: Crea
          *        to add the selected permissions to the created role.
          */
         createRole(roleData).then(response => {
-
+
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]:
```suggestion

```

File: apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx (hunk 1 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx]: fixed with commit 5fb1030

File: apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx (hunk 2 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-permisson.tsx]: fixed with commit 5fb1030

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx (hunk 3 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]: fixed with commit 5fb1030

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx (hunk 4 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]: fixed with commit 5fb1030

File: apps/admin-portal/src/components/roles/create-role-wizard/role-user-assign.tsx [hunk 5]
```diff
@@ -178,7 +169,9 @@ export const AddRoleUsers: FunctionComponent<AddUserRoleProps> = (props: AddUser
                                                         image={ user.profileUrl }
                                                     />
                                                      { user.userName }
-                                                     <Icon name="add" />
+                                                     <Icon
+                                                         name="add"
+                                                     />
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-user-assign.tsx]: Why break in to multiple lines ?

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx [hunk 6]
```diff
@@ -101,89 +101,16 @@ export const CreateRoleWizard: FunctionComponent<CreateRoleProps> = (props: Crea
...
      * @param permissions - permissions selected for the created role.
      */
     const addRole = (basicData: any): void => {
-        let members: CreateRoleMemberInterface[] = [];
-
-        if (basicData.RoleUserList.length > 0) {
-            basicData.RoleUserList.forEach(user => {
+        const members: CreateRoleMemberInterface[] = [];
+        if (basicData.RoleUserList.users.length > 0) {
+            basicData.RoleUserList.users.forEach(user => {
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]: Could assign `basicData.RoleUserList.users` to a `const` and use

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx [hunk 7]
```diff
@@ -193,7 +120,7 @@ export const CreateRoleWizard: FunctionComponent<CreateRoleProps> = (props: Crea
         
         const roleData: CreateRoleInterface = {
             "schemas": [
-                "urn:ietf:params:scim:schemas:core:2.0:Group"
+              "urn:ietf:params:scim:schemas:core:2.0:Group"
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]: Why make it 2 spaces ?

File: apps/admin-portal/src/components/roles/edit-role/edit-role-basic.tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    const { t } = useTranslation();
+    const dispatch = useDispatch();
+
+    const {
+        roleObject,
+        roleId
+    } = props;
+
+
+
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/edit-role/edit-role-basic.tsx]:
```suggestion

```

Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/edit-role/edit-role-basic.tsx]:
```suggestion

```

File: apps/admin-portal/src/components/roles/create-role-wizard/role-user-assign.tsx (hunk 5 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-user-assign.tsx]: Fixed with commit 3556a7e

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx (hunk 7 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]: fixed with commit f0e96ae

File: apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx (hunk 6 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/create-role-wizard/role-wizard.tsx]: fixed with commit f0e96ae

File: apps/admin-portal/src/components/roles/edit-role/edit-role-basic.tsx (hunk 8 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/edit-role/edit-role-basic.tsx]: fixed with commit f0e96ae

File: apps/admin-portal/src/components/roles/edit-role/edit-role-users.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import { AddRoleUsers } from "../create-role-wizard/role-user-assign"
+import { Forms } from "@wso2is/forms"
+import { Grid, Button } from "semantic-ui-react"
+
+export const RoleUserDetails: FunctionComponent<any> = (): ReactElement => {
+    return (
+        <>
+            <AddRoleUsers/>
+            <Forms onSubmit={() => {
+                console.log()
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/edit-role/edit-role-users.tsx]: Remove the console.log

File: apps/admin-portal/src/components/roles/role-list.tsx [hunk 10]
```diff
@@ -37,21 +38,29 @@ export const RoleList: React.FunctionComponent<RoleListProps> = (props: RoleList
         handleRoleDelete
     } = props;
 
+    const handleRoleEdit = (roleId: string) => {
+        history.push(`roles/${ roleId }`);
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/role-list.tsx]: Shall we get the `roles` path from a constant

File: apps/admin-portal/src/components/roles/role-utils.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        comp =  {
+            label: permObj.displayName,
+            fullPath: permObj.resourcePath,
+            isExpanded: true,
+            isChecked: false,
+            name: component
+        }
+        arr.push(comp)
+    }
+    if(pathcomponents.length){
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/role-utils.ts]:
```suggestion
    if (pathcomponents.length) {
```

File: apps/admin-portal/src/components/roles/role-utils.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+import { PermissionObject, Permission } from "../../models/permission"
+
+/**
+ * A Util method to create an array of permission object with heirarchy.
+ * 
+ * @param permObj - Permission Object
+ * @param pathcomponents - Permission path array
+ * @param arr - Empty array to start with
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/role-utils.ts]: What is this array is for. Shall we make it meaningful ?

File: apps/admin-portal/src/components/roles/role-utils.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            isExpanded: true,
+            isChecked: false,
+            name: component
+        }
+        arr.push(comp)
+    }
+    if(pathcomponents.length){
+        addPath(permObj, pathcomponents, comp.children || (comp.children = []))
+    }
+    return arr;
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/role-utils.ts]:
```suggestion

    return arr;
```

File: apps/admin-portal/src/components/roles/edit-role/edit-role-users.tsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+import { AddRoleUsers } from "../create-role-wizard/role-user-assign"
+import { Forms } from "@wso2is/forms"
+import { Grid, Button } from "semantic-ui-react"
+
+export const RoleUserDetails: FunctionComponent<any> = (): ReactElement => {
+    return (
+        <>
+            <AddRoleUsers/>
+            <Forms onSubmit={() => {
+                //TODO handle form submit
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/edit-role/edit-role-users.tsx]:
```suggestion
                // TODO: handle form submit
```

File: apps/admin-portal/src/pages/role-edit.tsx [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    const [ roleId, setRoleId ] = useState<string>(undefined);
+    const [ roleObject, setRoleObject ] = useState<RolesInterface>();
+
+    const getRoleDetails = (roleId: string ): void => {
+        getRoleById(roleId).then(response => {
+            if (response.status === 200) {
+                setRoleObject(response.data);
+            }
+        }).catch(error => {
+            //TODO handle error
```
Comment (jeradrutnam) [inline apps/admin-portal/src/pages/role-edit.tsx]:
```suggestion
            // TODO: handle error
```

File: apps/admin-portal/src/pages/role-edit.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    useEffect(() => {
+        const path = history.location.pathname.split("/");
+        const roleId = path[ path.length - 1 ];
+
+        setRoleId(roleId);
+        getRoleDetails(roleId);
+    }, []);
+
+    const handleBackButtonClick = () => {
+        history.push("/roles");
```
Comment (jeradrutnam) [inline apps/admin-portal/src/pages/role-edit.tsx]: Lets gets the path from a constant

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+        onUpdateCb(updatedData, depth);
+    }
+
+    const handleCheckToggle = (node: TreeNode, e: any) => {
+        const { onCheckToggleCb, depth } = props;
+        const data = _.cloneDeep(treeData);
+        const currentNode = _.find(data, node);
+        const currentNodeIndex = data.indexOf(currentNode);
+        const toggledNodes = [];
+        if (e.shiftKey && !_.isNil(lastCheckToggledNodeIndex)) {
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion

        if (e.shiftKey && !_.isNil(lastCheckToggledNodeIndex)) {
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+            const rangeStart = Math.min(
+                currentNodeIndex,
+                lastCheckToggledNodeIndex
+            );
+            const rangeEnd = Math.max(
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion

            const rangeEnd = Math.max(
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+            );
+        }
+    }
+
+    const printNoChildrenMessage = () => {
+        const {
+            transitionExitTimeout,
+            noChildrenAvailableMessage
+        } = props;
+        const noChildrenTransitionProps = {
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion

        const noChildrenTransitionProps = {
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+                                          'super-treeview-node' + getStyleClassCb(node)
+                                      }
+                                  >
+                                      <div className={`super-treeview-node-content ${!node.children 
+                                            || node.children.length == 0 ? "no-child" : ""}`}>
+                                          {node.children && node.children.length != 0 ? printExpandButton(node) : ""}
+                                          {printCheckbox(node)}
+                                          {printDeleteButton(node)}
+                                      </div>
+                                      {printChildren(node)}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                                      { printChildren(node) }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+                                  <div
+                                      className={
+                                          'super-treeview-node' + getStyleClassCb(node)
+                                      }
+                                  >
+                                      <div className={`super-treeview-node-content ${!node.children 
+                                            || node.children.length == 0 ? "no-child" : ""}`}>
+                                          {node.children && node.children.length != 0 ? printExpandButton(node) : ""}
+                                          {printCheckbox(node)}
+                                          {printDeleteButton(node)}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                                          { printCheckbox(node) }
                                          { printDeleteButton(node) }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 22]
```diff
@@ -0,0 +1,350 @@
...
+                    data={node[keywordChildren] || []}
+                    depth={depth + 1}
+                    onUpdateCb={onChildrenUpdateCb.bind(this)}
+                />
+            );
+        }
+
+        return (
+            <div className="super-treeview-children-container">
+                {childrenElement}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                { childrenElement }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+    keywordChildren: 'children',
+    keywordChildrenLoading: 'isChildrenLoading',
+    keywordLabel: 'name',
+    keywordKey: 'id',
+    loadingElement: <div>loading...</div>,
+    noChildrenAvailableMessage: 'No data found',
+    transitionEnterTimeout: 1200,
+    transitionExitTimeout: 1200,
+
+    getStyleClassCb: (/* node, depth */) => { return ''; },
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
    getStyleClassCb: (/* node, depth */) => { return ""; },
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+
+TreeView.defaultProps = {
+    depth: 0,
+    deleteElement: <div>(X)</div>,
+    keywordChildren: 'children',
+    keywordChildrenLoading: 'isChildrenLoading',
+    keywordLabel: 'name',
+    keywordKey: 'id',
+    loadingElement: <div>loading...</div>,
+    noChildrenAvailableMessage: 'No data found',
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]: Use double quotes

File: modules/theme/src/definitions/views/admin-portal.less [hunk 25]
```diff
@@ -115,14 +115,15 @@
 }
 
 // VARIABLES
-@super-treeview-text-color: #000;
-@super-treeview-arrow-color: #000;
-@super-treeview-tree-stem-color: #000;
+@super-treeview-tree-stem-color: #f3f3f3;
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: Move this to variables file

File: modules/theme/src/definitions/views/admin-portal.less [hunk 26]
```diff
@@ -115,14 +115,15 @@
 }
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]: Format the code

File: apps/admin-portal/src/components/roles/role-list.tsx (hunk 10 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/role-list.tsx]: Fixed

File: apps/admin-portal/src/components/roles/role-utils.ts (hunk 12 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/role-utils.ts]: Fixed

File: apps/admin-portal/src/components/roles/role-utils.ts (hunk 13 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/components/roles/role-utils.ts]: Fixed

File: apps/admin-portal/src/pages/role-edit.tsx (hunk 16 above)
Comment (DimalChandrasiri) [reply apps/admin-portal/src/pages/role-edit.tsx]: fixed

File: modules/theme/src/definitions/views/admin-portal.less (hunk 25 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: fixed

File: modules/theme/src/definitions/views/admin-portal.less (hunk 26 above)
Comment (DimalChandrasiri) [reply modules/theme/src/definitions/views/admin-portal.less]: Fixed

File: modules/react-components/src/tree-view/tree-view.tsx (hunk 24 above)
Comment (DimalChandrasiri) [reply modules/react-components/src/tree-view/tree-view.tsx]: fixed

File: apps/admin-portal/src/components/roles/role-utils.ts [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+ * @param permissionTreeArray - Empty array to start with
+ * 
+ * @returns {Permission[]} - Permission arra with tree structure
+ */
+export const addPath = (permObj: PermissionObject, pathcomponents: string[],
+        permissionTreeArray: Permission[]): Permission[] => {
+
+    const component = pathcomponents.shift()
+    let comp = permissionTreeArray.find(item => item.name === component)
+    if (!comp) {
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/role-utils.ts]:
```suggestion

    if (!comp) {
```

File: apps/admin-portal/src/components/roles/role-utils.ts [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+        comp = {
+            label: permObj.displayName,
+            fullPath: permObj.resourcePath,
+            isExpanded: true,
+            isChecked: false,
+            name: component
+        }
+        permissionTreeArray.push(comp)
+    }
+    if (pathcomponents.length) {
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/roles/role-utils.ts]:
```suggestion

    if (pathcomponents.length) {
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 29]
```diff
@@ -0,0 +1,356 @@
...
+                                  key={node[keywordKey] || index}
+                              >
+                                  <div
+                                      className={
+                                          "super-treeview-node" + getStyleClassCb(node)
+                                      }
+                                  >
+                                      <div className={`super-treeview-node-content ${!node.children 
+                                            || node.children.length == 0 ? "no-child" : ""}`}>
+                                          {node.children && node.children.length != 0 ? printExpandButton(node) : ""}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                                          { node.children && node.children.length != 0 ? printExpandButton(node) : "" }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 30]
```diff
@@ -0,0 +1,356 @@
...
+                                          {printCheckbox(node)}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                                          { printCheckbox(node) }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 31]
```diff
@@ -0,0 +1,356 @@
...
+                                          {printDeleteButton(node)}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                                          { printDeleteButton(node) }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 32]
```diff
@@ -0,0 +1,356 @@
...
+    const printCheckbox = (node: TreeNode) => {
+        const { isCheckable, keywordLabel, depth } = props;
+        const nodeText = _.get(node, keywordLabel, "");
+
+        if (isCheckable(node, depth)) {
+            return (
+                <label htmlFor={node.name} className="tree-label">
+                    <input
+                        type="checkbox"
+                        name={node[keywordLabel]}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                        name={ node[keywordLabel] }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 33]
```diff
@@ -0,0 +1,356 @@
...
+                        className="invisible"
+                        onClick={(e) => {
+                            handleCheckToggle(node, e);
+                        }}
+                        checked={!!node.isChecked}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                        checked={ !!node.isChecked }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 34]
```diff
@@ -0,0 +1,356 @@
...
+                        id={node.name}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                        id={ node.name }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 35]
```diff
@@ -0,0 +1,356 @@
...
+        const { isDeletable, depth, deleteElement } = props;
+
+        if (isDeletable(node, depth)) {
+            return (
+                <div className="delete-btn"
+                    onClick={() => {
+                        handleDelete(node);
+                    }}
+                >
+                    {deleteElement}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                    { deleteElement }
```

File: apps/admin-portal/src/pages/role-edit.tsx (hunk 16 above)
Comment (jeradrutnam) [reply apps/admin-portal/src/pages/role-edit.tsx]: Still it's the same

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 36]
```diff
@@ -0,0 +1,356 @@
...
+        handleUpdate(data);
+    }
+
+    const printCheckbox = (node: TreeNode) => {
+        const { isCheckable, keywordLabel, depth } = props;
+        const nodeText = _.get(node, keywordLabel, "");
+
+        if (isCheckable(node, depth)) {
+            return (
+                <label htmlFor={node.name} className="tree-label">
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                <label htmlFor={ node.name } className="tree-label">
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 37]
```diff
@@ -0,0 +1,356 @@
...
+        const { keywordChildren, keywordChildrenLoading, depth } = props;
+        const isChildrenLoading = _.get(node, keywordChildrenLoading, false);
+        let childrenElement;
+
+        if (isChildrenLoading) {
+            childrenElement = _.get(props, "loadingElement");
+        } else {
+            childrenElement = (
+                <TreeView
+                    {...props}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                    { ...props }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 38]
```diff
@@ -0,0 +1,356 @@
...
+                                            || node.children.length == 0 ? "no-child" : ""}`}>
+                                          { node.children && node.children.length != 0 ? printExpandButton(node) : "" }
+                                          { printCheckbox(node) }
+                                          { printDeleteButton(node) }
+                                      </div>
+                                      { printChildren(node) }
+                                  </div>
+                              </CSSTransition>
+                          );
+                      })}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                { _.isEmpty(nodeArray)
                    ? printNoChildrenMessage()
                    : nodeArray.map((node, index) => {
                          return (
                              <CSSTransition
                                  { ...nodeTransitionProps }
                                  key={ node[keywordKey] || index }
                              >
                                  <div
                                      className={
                                          "super-treeview-node" + getStyleClassCb(node)
                                      }
                                  >
                                      <div className={ `super-treeview-node-content ${ !node.children 
                                            || node.children.length == 0 ? "no-child" : "" }` }>
                                          { node.children && node.children.length != 0 ? printExpandButton(node) : "" }
                                          { printCheckbox(node) }
                                          { printDeleteButton(node) }
                                      </div>
                                      { printChildren(node) }
                                  </div>
                              </CSSTransition>
                          );
                      }) }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 39]
```diff
@@ -0,0 +1,356 @@
...
+            currentNode[keywordChildren] = updatedData;
+            handleUpdate(data);
+        }
+    }
+
+    return (
+        //TODO: The extra container will be removed when styling is fixed.
+        <div className="tree-vew">
+            <div className="super-treeview">
+                {printNodes(treeData)}
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
                { printNodes(treeData) }
```

File: modules/react-components/src/tree-view/tree-view.tsx [hunk 40]
```diff
@@ -0,0 +1,356 @@
...
+            const data = _.cloneDeep(treeData);
+            const currentNode = _.find(data, node);
+
+            currentNode[keywordChildren] = updatedData;
+            handleUpdate(data);
+        }
+    }
+
+    return (
+        //TODO: The extra container will be removed when styling is fixed.
```
Comment (jeradrutnam) [inline modules/react-components/src/tree-view/tree-view.tsx]:
```suggestion
        // TODO: The extra container will be removed when styling is fixed.
```

File: modules/theme/src/definitions/views/admin-portal.less [hunk 41]
```diff
@@ -215,175 +215,176 @@ TODO : Revist the styling in the component
...
-                path {
-                    fill: @checkboxfill;
+    
+    .tree-label {
+        cursor: pointer;
+        margin-bottom: 0;
+        user-select: none;
+    
+        input {
+
```
Comment (jeradrutnam) [inline modules/theme/src/definitions/views/admin-portal.less]:
```suggestion
```

**Pull request merged**